File: leetcode/032_Longest_Valid_Parentheses/LongestValidParentheses.cpp
#include <cstdio>
#include <cstdlib>
#include <iostream>
#include <stack>
#include <string>

using namespace std;

class Solution
{
  public:
    int longestValidParentheses(string s)
    {
        int res = 0, start = 0;
        stack<int> st;
        for (int i = 0; i < s.length(); i++)
        {
            if (s[i] == '(')
            {
                st.push(i);
            }
            else
            {
                if (st.empty())
                {
                    start = i + 1;
                }
                else
                {
                    st.pop();
                    res = st.empty() ? max(res, i - start + 1) : max(res, i - st.top());
                }
            }
        }
        return res;
    }
};

int main()
{
    Solution so;
    int result;
    result = so.longestValidParentheses(")()");
    cout << result << endl;
    result = so.longestValidParentheses(")()())");
    cout << result << endl;
    result = so.longestValidParentheses(")(");
    cout << result << endl;
    result = so.longestValidParentheses("(()");
    cout << result << endl;
    result = so.longestValidParentheses("()(()");
    cout << result << endl;
    result = so.longestValidParentheses(")(((((()())()()))()(()))(");
    cout << result << endl;

    system("pause");
    return 0;
}

File: leetcode/032_Longest_Valid_Parentheses/go/longestValidParentheses.go
package main

func longestValidParentheses(s string) int {
	var stack []int
	var max int
	stack = append(stack, -1)
	for i := 0; i < len(s); i++ {
		if s[i] == '(' {
			stack = append(stack, i)
		} else { // ')'
			stack = stack[:len(stack)-1]
			if len(stack) == 0 {
				stack = append(stack, i)
			} else {
				if i-stack[len(stack)-1] > max {
					max = i - stack[len(stack)-1]
				}
			}
		}
	}
	return max
}
